#GSOC PR: Enhance CNN Model with L2 Regularization, K-Fold CV, and Ensemble Techniques

sambhavnoobcoder opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- Please select appropriate Priority, Status,and Type labels-->
<!--- If you do not have permission to select labels please state which labels you would like -->

## Description
<!--- Describe your changes in detail -->
This PR enhances the CNN model in the SDA_downscale function with L2 Regularization, K-Fold Cross-Validation, and Ensemble techniques. Key improvements include:

- L2 Regularization added to dense layers to prevent overfitting.
- K-Fold Cross-Validation (k=5) implemented for robust model evaluation.
- Ensemble methods introduced with bagging (5 models per fold).
- CNN architecture refined with dual batch normalization and dropout layers.
- Enhanced performance evaluation with per-fold metrics and final test set evaluation.

These changes aim to improve model robustness, reduce overfitting, and increase prediction accuracy for environmental data downscaling.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
These enhancements address common challenges in machine learning, particularly for complex environmental data. They provide a more reliable framework for downscaling, improving generalization to unseen data and stability in training.

## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [x] Immediately
- [ ] Within one week
- [ ] When possible

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] My name is in the list of CITATION.cff
- [ ] I have updated the CHANGELOG.md.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

sambhavnoobcoder committed:
Update downscale_function.R

Add L2 regularization to CNN model in SDA_downscale function

- Introduced L2 kernel regularization to convolutional and dense layers
- Set L2 factor to 0.01 for all regularized layers
- Aim to reduce overfitting and improve model generalization
- Updated model architecture in the CNN branch of SDA_downscale function

sambhavnoobcoder committed:
Define k-folds and create cross-validation indices

- Set k_folds variable to 5 for k-fold cross-validation
- Use caret::createFolds to generate fold indices for training data

sambhavnoobcoder committed:
Implement cross-validation loop for ensemble modeling

- Initialize cv_results list to store results
- Create loop to process each fold
- Split data into training and validation sets for each fold
- Print progress message for each fold

sambhavnoobcoder committed:
Implement model training and evaluation within cross-validation

- Use tryCatch for error handling during model fitting
- Evaluate model on validation set
- Store results in cv_results list
- Handle potential errors and log them

sambhavnoobcoder committed:
Calculate and display average cross-validation performance

- Compute mean MSE and MAE across all folds
- Use sapply for efficient calculation
- Handle potential NA values
- Print results for the current ensemble

sambhavnoobcoder committed:
Implement bagging for CNN in SDA_downscale function

- Add num_bags parameter to control number of bagged models
- Create and train multiple CNN models for each ensemble
- Implement bagged prediction function for CNN
- Update prediction process to use bagged models
- Adjust cross-validation to incorporate bagging

sambhavnoobcoder committed:
Improve bagged CNN evaluation and standardize nomenclature

- Add final test set evaluation for each bagged CNN ensemble
- Calculate and display Test MSE and MAE for ensembles
- Standardize variable names for clarity and suiting the flow of the code avoiding conflicts .

sambhavnoobcoder committed:
Added number of bags for bagging 

- instead of passing as function arg , passed inside function 
- set the number of bags to 5

sambhavnoobcoder committed:
Refactor CNN model with k-fold cross-validation and bagging

- Implement k-fold cross-validation for CNN models
- Add bagging to improve model robustness
- Restructure model creation and training process
- Introduce ensemble prediction function
- Update evaluation metrics for each fold and final ensemble
- Improve code organization and readability

sambhavnoobcoder committed:
Added comments 

some snippets require comments to explain their purpose , they have been added as well , as well as some past comments have been modified to aptly suit the changes .

sambhavnoobcoder committed:
Removed MAE ,MSE from downscale 

- MAE , MSE already in metrics function
- reimplementing it makes it redundant 
- Removed redundant implementation from downscale function

mdietze commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -140,84 +140,141 @@ SDA_downscale <- function(preprocessed, date, carbon_pool, covariates, model_typ
+        
+        # Create bagged models for this fold
+        fold_models <- list()
+        for (bag in 1:num_bags) {
```
Unless I'm misunderstanding how things are done below, this looks like a place where the folds could in theory be done in parallel, rather than sequentially. You might ping @DongchenZ as he'd been looking closely recently at the different parallelization libraries in R and their relative efficiency and memory overhead.

mdietze commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -140,84 +140,141 @@ SDA_downscale <- function(preprocessed, date, carbon_pool, covariates, model_typ
-        decay_rate = 0.9
-      )
+      # Create k-fold indices
+      fold_indices <- caret::createFolds(y = 1:nrow(x_train), k = k_folds, list = TRUE, returnTrain = FALSE)
```
This introduces a new dependency for what is actually appears to be a fairly straightforward task. Are there analogous functions you could use from packages that are already in the dependencies? Could you do this manually with base functions like `sample` and/or a bit of tidyverse?  If the answers to this are no, you need to add the new package to the dependencies and update the PR.

dlebauer replied:
FWIW the 'tidyverse' equivalent would be `rsample::vfold_cv`, but still might not be worth adding this dependency since implementation "should be" 🤞🏻 simple, like:

```
# <incomplete function documentation>
# @param i vector of indices created by seq_along(x)
# @param k number of groups
# based on https://stackoverflow.com/a/57085061/199217
create_folds <- function(i, k) {
  indices <- sample(i)
  folds <- split(indices, cut(indices, breaks = k, labels = FALSE))
  return(folds)
}
```

Ok it wasn't as simple as I hoped, but that should work!

sambhavnoobcoder replied:
Addressed in the commits [4ae1974](https://github.com/PecanProject/pecan/pull/3357/commits/4ae1974d7aa78fa27e188d222e85ab0eba14b773) , [5d86312](https://github.com/PecanProject/pecan/pull/3357/commits/5d86312d92fe03adeb3a77665b96db4ec2dcda45) and [70bfdaf](https://github.com/PecanProject/pecan/pull/3357/commits/70bfdaf5ac40a69309d78b063c28d41576d5603d)

dlebauer commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -140,84 +140,141 @@ SDA_downscale <- function(preprocessed, date, carbon_pool, covariates, model_typ
+        cat(sprintf("Processing ensemble %d, fold %d of %d\n", i, fold, k_folds))
+        
+        # Split data into training and validation sets for this fold
+        train_indices <- setdiff(1:nrow(x_train), fold_indices[[fold]])
```
to improve code robustness, replace uses of `1:length(x)` with `seq_along(x)`

sambhavnoobcoder replied:
Addressed in commit [bf02478](https://github.com/PecanProject/pecan/pull/3357/commits/bf024780296e8c05680ad19b9abaa3d61c0b4611)

dlebauer commented on modules/assim.sequential/R/downscale_function.R:
```diff
@@ -140,84 +140,141 @@ SDA_downscale <- function(preprocessed, date, carbon_pool, covariates, model_typ
-      )
+      #initialise operations for each fold
+      for (fold in 1:k_folds) {
+        cat(sprintf("Processing ensemble %d, fold %d of %d\n", i, fold, k_folds))
```
I like the idea of providing updates on where we are. But do we want perhaps 100s of messages? This might be a good place to use the progress bar instead (see e.g. https://github.com/PecanProject/pecan/blob/325d7fc32fd38ebe53ce5ee81262dc307202daf5/modules/emulator/R/predict.GP.R#L103) 

Still it is a great idea to provide messages that contain useful information about the progress. If that is the case here, I think using the PEcAn logger functions would be preferred to `cat`, e.g.

```
logger.info(sprintf("Processing ensemble %d, fold %d of %d", i, fold, k_folds))
```

(though as mentioned above, this might generate too many messages to be useful, could just print once per fold? And if there is some useful information that could be provided when processing of each fold, perhaps add another logger.info when the fold is finished?

dlebauer reviewed (commented):
Left a few comments as an initial review, mostly on code style.

sambhavnoobcoder committed:
Merge branch 'PecanProject:develop' into CNN-Advance-2

sambhavnoobcoder committed:
Improve robustness of train-validation split in SDA_downscale function

- Replace 1:nrow(x_train) with seq_len(nrow(x_train)) for indexing
- Ensures proper handling of edge cases, including empty datasets
- Addresses potential "subscript out of bounds" error in CNN model training

sambhavnoobcoder committed:
Implement custom create_folds function to replace caret dependency

- Add create_folds function for k-fold cross-validation
- Removes dependency on caret::createFolds
- Supports both training and test set index generation
- Allows flexible output as list or vector

sambhavnoobcoder committed:
Add comprehensive roxygen documentation for create_folds function

- Include title, name, and author tags
- Detail all function parameters
- Provide clear description and details sections
- Specify return value format
- Align documentation style with project standards

sambhavnoobcoder committed:
Replace carets

- Replace caretFolds -> caret_folds
- Remove carets:: dependency
